go-controller/pkg/network-controller-manager/netattachdefcontroller.go

```diff
@@ -0,0 +1,544 @@
+	NewDummyNetworkController(topologyType, netName string) (NetworkController, error)
+}
+
+type netAttachDefinitionController struct {
+	ncm                NetworkControllerManager
+	nadFactory         networkattachmentdefinitioninformerfactory.SharedInformerFactory
+	netAttachDefLister networkattachmentdefinitionlisters.NetworkAttachmentDefinitionLister
+	netAttachDefSynced cache.InformerSynced
+	queue              workqueue.RateLimitingInterface
+	loopPeriod         time.Duration
+
+	NetAttachDefManager
+}
+
+func NewNadController(ncm NetworkControllerManager, ovnClientset *util.OVNClientset) *netAttachDefinitionController {
+	nadFactory := netattachdefinformers.NewSharedInformerFactoryWithOptions(
+		ovnClientset.NetworkAttchDefClient,
+		avoidResync,
+	)
+	netAttachDefInformer := nadFactory.K8sCniCncfIo().V1().NetworkAttachmentDefinitions()
+	rateLimter := workqueue.NewMaxOfRateLimiter(
+		workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second),
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(qps), qps*5)})
+
+	nadController := &netAttachDefinitionController{
+		ncm:                ncm,
+		nadFactory:         nadFactory,
+		netAttachDefLister: netAttachDefInformer.Lister(),
+		netAttachDefSynced: netAttachDefInformer.Informer().HasSynced,
+		queue:              workqueue.NewNamedRateLimitingQueue(rateLimter, "net-attach-def"),
+		loopPeriod:         time.Second,
+
+		NetAttachDefManager: &netAttachDefManager{
+			perNadNetConfInfo:     syncmap.NewSyncMap[*nadNetConfInfo](),
+			perNetworkNadNameInfo: syncmap.NewSyncMap[*nadNameInfo](),
+		},
+	}
+	netAttachDefInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    nadController.onNetworkAttachDefinitionAdd,
+			UpdateFunc: nadController.onNetworkAttachDefinitionUpdate,
+			DeleteFunc: nadController.onNetworkAttachDefinitionDelete,
+		})
+	return nadController
+}
+
+func (nadController *netAttachDefinitionController) Run(stopChan <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer nadController.queue.ShutDown()
+
+	nadController.nadFactory.Start(stopChan)
+	klog.Infof("Starting controller %s", controllerName)
+	if !cache.WaitForNamedCacheSync(controllerName, stopChan, nadController.netAttachDefSynced) {
+		return fmt.Errorf("error syncing cache")
+	}
+
+	err := nadController.repairNads()
+	if err != nil {
+		klog.Errorf("Failed to sync all existing nad entries: %v", err)
+	}
+
+	klog.Info("Starting workers for controller %s", controllerName)
+	wg := &sync.WaitGroup{}
+	for i := 0; i < numberOfWorkers; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			wait.Until(nadController.worker, nadController.loopPeriod, stopChan)
+		}()
+	}
+	wg.Wait()
+
+	// wait until we're told to stop
+	<-stopChan
+
+	klog.Infof("Shutting down controller %s", controllerName)
+	nadController.queue.ShutDown()
+	return nil
+}
+
+func (nadController *netAttachDefinitionController) repairNads() (err error) {
```

Comment:
not sure I understand your comments, SyncNetworkControllers() is the entry point for different controller managers.

Reply:
I guess I am confused about these things:
1. Naming this method `repairNads` gives a feeling that we are changing the NADs, but we are not. `SyncNetworkControllers` seems the more appropriate name for this method as well actually.
2. `GetAllNetworkControllers` could be called from within `SyncNetworkControllers`
3. The current name `SyncNetworkControllers` is ok, but could be named better `CleanupDeletedNetworks` or such...
4. `NewDummyNetworkController` from the `NetworkControllerManager` interface could be removed since it is only ever used within `NetworkControllerManager` and not from outside.